rfcs/build-system/01_convergence_devstack.md

```diff
@@ -0,0 +1,191 @@
+
+**👉 tsconfig path aliases + TS solution style configs**
+
+Rather then leveraging solutions that might feel appealing (like TS Project refs etc, lerna/yarn workspace symlinks), we can leverage battle tested solution that has been in place for quite some time, thus makes its adoption compatibility with 3rd party toolign straightforward.
+
+**Pre-requirements**
+
+- typescript 3.9
+
+**Example:**
+
+> @file /tsconfig.base.json
+> "root tsconfig" (intentionally not named as tsconfig.json, so editor/IDE will pick solution config per package)
+
+```json
+{
+  "extends": "@fluentui/ts-config-common",
+  "compilerOptions": {
+    "lib": ["dom", "es2017", "dom.iterable"],
+    "rootDir": ".",
+    "baseUrl": ".",
+    "paths": {
+      "@fluentui/react-button": ["packages/react-button/src/index.ts"],
+      "@fluentui/react-hooks": ["packages/react-hooks/src/index.ts"]
+      // other aliases...
+    }
+  }
+}
+```
+
+> @file /packages/react-button/tsconfig.json
+> root package tsconfig - [solution config style](https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/#solution-style-tsconfig)
+
+```json
+{
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "types": ["node", "jest"]
+  },
+  "include": [],
+  "files": [],
+  "references": [
+    {
+      // lib type config is used for all implementation files (ambient types generation)
+      "path": "./tsconfig.lib.json"
+    },
+    {
+      // spec type config is used for tests only, thus mitigates leaking globals into implementation etc. This is not used for prod builds
+      "path": "./tsconfig.spec.json"
+    }
+  ]
+}
+```
+
+![solution TS config per environment resolution](https://i.imgur.com/U58q2Ed.png)
+
+- solution config per package enables us to author different tsconfigs per environment (unit test env -> `tsconfig.spec.json` / implementation env -> `tsconfig.lib.json`), thus properly scoping type checker and other benefits.
+- With `tsconfig.base.json` in place, we can leverage common oss tooling like [tsconfig-paths package](https://github.com/dividab/tsconfig-paths#readme) to leverage path aliases config within all other tools that we (may) use (jest,webpack,rollup,storybook...).
+- any code change will instantly propagate to whole TS tree and tools that we use (no physicall rebuilds needed - things will change in memory)
+
+#### 2. unified TSConfigs
+
+Current state:
+
+- N\* (TBA)
+- OUIF (TBA)
+
+Convergence proposal:
+
+- based on #1 we can easily reflect TS solution style configs generation (lib/spec) within solid scaffolding tool.
+
+**👉 common scaffolding tool**
+
+Currently we are using Plop which is good enough for now.
+
+Going forward I'd suggest to go with more robust well known tooling such as Angular Schematics/[Nx Devkit generators](https://nx.dev/latest/react/core-concepts/nx-devkit#generators) with various features:
```

Comment:
What's the ROI on replacing plop? Is creating new packages a regular enough occurrence that we'd save more time than it would take to build on top of a new engine? (Also there's a question of education; I've created many packages in this repo and not known about plop, but had manually created packages instead. It never takes more than a few minutes, so it's not immediately clear that a scaffolding tool is necessary at all.)

Reply:
@jdhuntington Creating packages will be much more common in the next couple months as we make more converged components, and having good tooling (whether based on plop or nx) is important because often the people creating the packages won't have as good a grasp on the tooling/configuration aspects as you or I do. The tooling also helps to ensure that configs and such use current patterns.